File: code/mathematical-algorithms/next_larger_number/next_larger_number.cpp
#include<bits/stdc++.h>
using namespace std;
// Part of Cosmos by OpenGenus Foundation
void nextgr(char num[],int n)
{
 int i,j;
 
 for(i=n-1;i>=0;i--)
 {
   if(num[i-1]< num[i])
   break;
 }
 
 if(i==0)
 {
  cout<<"No such greater number exists";
  return ;
 }
 
 int x=num[i-1], small=i;
 
 for(j=i+1;j<n;j++)
 {
   if(num[j] > x && num[j] < num[small] )
   small=j;
 }
 
 swap(num[i-1],num[small]);
 
 sort(num+i, num+n);
 
 cout<<num;
}   

int main()
{
 char num[20];
 
 cin>>num;
 int n=strlen(num);
 nextgr(num,n);
 
 cout<<endl;
 return 0;
}
       

File: code/mathematical-algorithms/next_larger_number/next_larger_number.java
// Part of Cosmos by OpenGenus Foundation
public class NextGreaterNumber 
{
    private void swap(int[] number, int i, int j)
    {
        int temp  = number[i];
        number[i] = number[j];
        number[j] = temp;
    }
     
    private void sortSubarray(int[] number, int i, int j)
    {
        
        while (i < j)
        {
            swap(number, i, j);
            i += 1;
            j -= 1;
        }
    }
     
    public void findNextGreaterNumber(int[] number)
    {
        int lastDigitSeen = number[number.length-1], i, j;
        for (i = number.length-2; i >= 0; i--)
        {
           
            if (lastDigitSeen > number[i])
            {
                break;
            }
            lastDigitSeen = number[i];
        }
         
        if (i >= 0) // we found the digit breaking the sorted ordering
        {
           
            for (j = number.length-1; j > i; j--)
            {
                if (number[j] > number[i])
                {
                    break;
                }
            }
             
           
            swap(number, i, j);
             
            
            sortSubarray(number, i+1, number.length-1);
        }
    }
     
     
    public static void main(String[] args) 
    {
        NextGreaterNumber solution = new NextGreaterNumber();
         
        int[] number = {6,9,3,8,6,5,2};
         
        solution.findNextGreaterNumber(number);
         
        System.out.println("Next greater number is: ");
        for (int i = 0; i < number.length; i++)
        {
            System.out.print(number[i]);
        }
    }
}
